File: jangwon/Self/알고리즘구현/다익스트라 문제/배달.py
import heapq
import collections

def solution(N, road, K):
    graph = [[] for _ in range(N + 1)]
    
    for u, v, w in road:
        graph[u].append((v, w))
        graph[v].append((u, w))

    # 시작 노드, 소요 시간
    Q = [(1, 0)]
    dist = collections.defaultdict(int)
    
    while Q:
        node, time = heapq.heappop(Q)
        if node not in dist:
            dist[node] = time
            for v, w in graph[node]:
                alt = time  + w
                heapq.heappush(Q, (v, alt))
                
    result = 0
    for v in dist.values():
        if v <= K:
            result += 1
            
    return result

File: jangwon/Self/알고리즘구현/다익스트라 문제/배달.js
// 프로그래머스 배달
// 다익스트라로 해결
function solution(N, road, K) {
    const visited = Array.from({length: N + 1}, () => Infinity);
    const graph = Array.from({length: N + 1}, () => Array());
    let result = 0;
    
    // 그래프 구성
    for (const [u, v, w] of road) {
        graph[u].push([v, w]);
        graph[v].push([u, w]);
    }
    
    // 시작 노드, 배달 시간
    let Q = [[1, 0]];
    visited[1] = 0;
    
    while (Q.length) {
        const [node, time] = Q.shift();
        
        for (let i = 0; i < graph[node].length; i++) {
            // 최소 값 갱신
            const next = graph[node][i];
            if (visited[next[0]] > visited[node] + next[1]) {
                visited[next[0]] = visited[node] + next[1];
                Q.push(next);
            }
        }
    }
    
    visited.forEach((v) => (v <= K ? result += 1 : false));
    return result;
}